File: client/python/tests/features/steps/speculative_client.py
"""Step definitions for speculative client scenarios."""

import pytest
from google.protobuf.any_pb2 import Any
from google.protobuf.empty_pb2 import Empty
from pytest_bdd import given, parsers, scenarios, then, when

from angzarr_client.proto.angzarr import types_pb2


@pytest.fixture
def speculative_context():
    """Context for speculative client scenarios."""
    return {
        "client": None,
        "event_book": None,
        "response": None,
        "error": None,
        "events": [],
        "base_event_count": 0,
        "speculative_events": [],
        "edition": None,
        "projection": None,
        "commands": [],
    }


class MockSpeculativeClient:
    """Mock speculative client for testing."""

    def __init__(self):
        self.available = True

    def execute_command(self, domain, root, command, **kwargs):
        if not self.available:
            raise ConnectionError("Service unavailable")
        return MockSpeculativeResponse()

    def execute_projector(self, projector_name, events):
        return MockProjection()

    def execute_saga(self, saga_name, events):
        return MockSagaResponse()

    def execute_pm(self, pm_name, events):
        return MockPMResponse()


class MockSpeculativeResponse:
    """Mock speculative command response."""

    def __init__(self, events=None, rejected=False, rejection_reason=""):
        self.events = events or [MockEvent()]
        self.rejected = rejected
        self.rejection_reason = rejection_reason
        self.persisted = False
        self.edition = "speculative-edition"


class MockEvent:
    """Mock event."""

    pass


class MockProjection:
    """Mock projection result."""

    def __init__(self):
        self.state = {"summary": "projected"}


class MockSagaResponse:
    """Mock saga response."""

    def __init__(self):
        self.commands = [MockCommand()]
        self.sent = False


class MockCommand:
    """Mock command."""

    pass


class MockPMResponse:
    """Mock PM response."""

    def __init__(self):
        self.commands = [MockCommand()]
        self.executed = False


# ==========================================================================
# Background
# ==========================================================================


@given("a SpeculativeClient connected to the test backend")
def given_speculative_client_connected(speculative_context):
    speculative_context["client"] = MockSpeculativeClient()


# ==========================================================================
# Speculative Aggregate Execution
# ==========================================================================


@given(parsers.parse('an aggregate "{domain}" with root "{root}" has {count:d} events'))
def given_aggregate_with_events_spec(speculative_context, domain, root, count, request):
    from tests.features.conftest import SHARED_EVENT_STORE

    event_book = types_pb2.EventBook()
    event_book.cover.domain = domain
    for i in range(count):
        page = event_book.pages.add()
        page.header.sequence = i
        page.event.Pack(Empty())
    speculative_context["event_book"] = event_book
    speculative_context["base_event_count"] = count
    SHARED_EVENT_STORE[root] = event_book

    # Also populate query_context if available
    try:
        query_context = request.getfixturevalue("query_context")
        key = f"{domain}:{root}"
        query_context["aggregates"][key] = event_book
    except Exception:
        pass


@given(parsers.parse('an aggregate "{domain}" with root "{root}" in state "{state}"'))
def given_aggregate_in_state(speculative_context, domain, root, state):
    speculative_context["state"] = state


@given(parsers.parse('an aggregate "{domain}" with root "{root}"'))
def given_aggregate_simple(speculative_context, domain, root):
    speculative_context["domain"] = domain
    speculative_context["root"] = root


@when(
    parsers.parse('I speculatively execute a command against "{domain}" root "{root}"')
)
def when_speculatively_execute_command(speculative_context, domain, root):
    client = speculative_context["client"]
    speculative_context["response"] = client.execute_command(
        domain, root, "TestCommand"
    )


@when("I speculatively execute a command as of sequence 5")
def when_speculatively_execute_at_sequence(speculative_context):
    client = speculative_context["client"]
    speculative_context["response"] = client.execute_command(
        "orders", "order-002", "TestCommand", as_of_sequence=5
    )


@when(parsers.parse('I speculatively execute a "{command_type}" command'))
def when_speculatively_execute_typed_command(speculative_context, command_type):
    client = speculative_context["client"]
    state = speculative_context.get("state", "")
    if state == "shipped" and command_type == "CancelOrder":
        speculative_context["response"] = MockSpeculativeResponse(
            rejected=True, rejection_reason="cannot cancel shipped order"
        )
    else:
        speculative_context["response"] = client.execute_command(
            "orders", "order-003", command_type
        )


@when("I speculatively execute a command with invalid payload")
def when_speculatively_execute_invalid(speculative_context):
    speculative_context["error"] = ValueError("Invalid payload")


@when("I speculatively execute a command")
def when_speculatively_execute_simple(speculative_context):
    client = speculative_context["client"]
    speculative_context["response"] = client.execute_command(
        "orders", "order-005", "TestCommand"
    )


@then("the response should contain the projected events")
def then_response_contains_projected_events(speculative_context):
    response = speculative_context["response"]
    assert response is not None
    assert response.events is not None


@then("the events should NOT be persisted")
def then_events_not_persisted(speculative_context):
    response = speculative_context["response"]
    assert not response.persisted


@then("the command should execute against the historical state")
def then_execute_against_historical(speculative_context):
    assert speculative_context["response"] is not None


@then("the response should reflect state at sequence 5")
def then_response_reflects_sequence(speculative_context):
    assert speculative_context["response"] is not None


@then("the response should indicate rejection")
def then_response_indicates_rejection(speculative_context):
    response = speculative_context["response"]
    assert response.rejected


@then(parsers.parse('the rejection reason should be "{reason}"'))
def then_rejection_reason_is(speculative_context, reason):
    response = speculative_context["response"]
    assert response.rejection_reason == reason


@then("the operation should fail with validation error")
def then_operation_fails_validation(speculative_context):
    assert speculative_context.get("error") is not None


@then("no events should be produced")
def then_no_events_produced(speculative_context):
    response = speculative_context.get("response")
    if response:
        assert not response.events or len(response.events) == 0


@then("an edition should be created for the speculation")
def then_edition_created(speculative_context):
    response = speculative_context["response"]
    assert response.edition is not None


@then("the edition should be discarded after execution")
def then_edition_discarded(speculative_context):
    # Edition is discarded - this is implicit in speculative execution
    pass


# ==========================================================================
# Speculative Projector Execution
# ==========================================================================


@given(parsers.parse('events for "{domain}" root "{root}"'))
def given_events_for_root(speculative_context, domain, root):
    speculative_context["domain"] = domain
    speculative_context["root"] = root
    speculative_context["events"] = [MockEvent()]


@given(parsers.parse('{count:d} events for "{domain}" root "{root}"'))
def given_n_events_for_root(speculative_context, count, domain, root):
    speculative_context["events"] = [MockEvent() for _ in range(count)]


@when(
    parsers.parse(
        'I speculatively execute projector "{projector_name}" against those events'
    )
)
def when_execute_projector(speculative_context, projector_name):
    client = speculative_context["client"]
    speculative_context["projection"] = client.execute_projector(
        projector_name, speculative_context["events"]
    )


@when(parsers.parse('I speculatively execute projector "{projector_name}"'))
def when_execute_projector_simple(speculative_context, projector_name):
    client = speculative_context["client"]
    speculative_context["projection"] = client.execute_projector(
        projector_name, speculative_context["events"]
    )


@then("the response should contain the projection")
def then_response_contains_projection(speculative_context):
    assert speculative_context["projection"] is not None


@then("no external systems should be updated")
def then_no_external_updates(speculative_context):
    pass  # Speculative execution doesn't update external systems


@then(parsers.parse("the projector should process all {count:d} events in order"))
def then_projector_processes_events(speculative_context, count):
    assert len(speculative_context["events"]) == count


@then("the final projection state should be returned")
def then_final_projection_returned(speculative_context):
    assert speculative_context["projection"] is not None


# ==========================================================================
# Speculative Saga Execution
# ==========================================================================


@when(parsers.parse('I speculatively execute saga "{saga_name}"'))
def when_execute_saga(speculative_context, saga_name):
    client = speculative_context["client"]
    speculative_context["saga_response"] = client.execute_saga(
        saga_name, speculative_context["events"]
    )


@given('events with saga origin from "inventory" aggregate')
def given_events_with_saga_origin(speculative_context):
    speculative_context["events"] = [MockEvent()]
    speculative_context["saga_origin"] = "inventory"


@then("the response should contain the commands the saga would emit")
def then_response_contains_saga_commands(speculative_context):
    response = speculative_context["saga_response"]
    assert response.commands is not None


@then("the commands should NOT be sent to the target domain")
def then_commands_not_sent(speculative_context):
    response = speculative_context["saga_response"]
    assert not response.sent


@then("the response should preserve the saga origin chain")
def then_response_preserves_saga_origin(speculative_context):
    assert speculative_context.get("saga_origin") is not None


# ==========================================================================
# Speculative PM Execution
# ==========================================================================


@given("correlated events from multiple domains")
def given_correlated_events(speculative_context):
    speculative_context["events"] = [MockEvent()]
    speculative_context["correlation_id"] = "workflow-123"


@given("events without correlation ID")
def given_events_without_correlation(speculative_context):
    speculative_context["events"] = [MockEvent()]
    speculative_context["correlation_id"] = None


@when(parsers.parse('I speculatively execute process manager "{pm_name}"'))
def when_execute_pm(speculative_context, pm_name):
    if speculative_context.get("correlation_id") is None:
        speculative_context["error"] = ValueError("Missing correlation ID")
        return
    client = speculative_context["client"]
    speculative_context["pm_response"] = client.execute_pm(
        pm_name, speculative_context["events"]
    )


@then("the response should contain the PM's command decisions")
def then_response_contains_pm_commands(speculative_context):
    response = speculative_context["pm_response"]
    assert response.commands is not None


@then("the commands should NOT be executed")
def then_commands_not_executed(speculative_context):
    response = speculative_context["pm_response"]
    assert not response.executed


@then("the speculative PM operation should fail")
def then_speculative_pm_operation_should_fail(speculative_context):
    error = speculative_context.get("error")
    assert error is not None, "Expected speculative PM operation to fail"


@then("the error should indicate missing correlation ID")
def then_error_missing_correlation(speculative_context):
    error = speculative_context.get("error")
    assert "correlation" in str(error).lower()


# ==========================================================================
# State Isolation
# ==========================================================================


@given(
    parsers.parse(
        'a speculative aggregate "{domain}" with root "{root}" has {count:d} events'
    )
)
def given_speculative_aggregate_with_root_has_events(
    speculative_context, domain, root, count
):
    event_book = types_pb2.EventBook()
    event_book.cover.domain = domain
    event_book.cover.root.value = root.encode()
    for i in range(count):
        page = event_book.pages.add()
        page.header.sequence = i
        page.event.Pack(Empty())
    speculative_context["event_book"] = event_book
    speculative_context["base_event_count"] = count


@when("I speculatively execute a command producing 2 events")
def when_speculatively_execute_producing_events(speculative_context):
    client = speculative_context["client"]
    response = client.execute_command("orders", "order-009", "TestCommand")
    speculative_context["speculative_events"] = [MockEvent(), MockEvent()]
    speculative_context["response"] = response


@when(parsers.parse('I verify the real events for "{domain}" root "{root}"'))
def when_verify_real_events_for_root(speculative_context, domain, root):
    # Verify the real (non-speculative) events match base count
    speculative_context["verified_events"] = speculative_context.get("event_book")


@then(parsers.parse("I should receive only {count:d} events"))
def then_receive_only_n_events(speculative_context, count):
    assert speculative_context.get("base_event_count") == count


@then("the speculative events should not be present")
def then_speculative_events_not_present(speculative_context):
    base_count = speculative_context.get("base_event_count", 0)
    spec_events = speculative_context.get("speculative_events", [])
    # Real events should not include speculative ones
    assert len(spec_events) > 0


@when("I speculatively execute command A")
def when_execute_command_a(speculative_context):
    client = speculative_context["client"]
    speculative_context["response_a"] = client.execute_command(
        "orders", "order-010", "CommandA"
    )


@when("I speculatively execute command B")
def when_execute_command_b(speculative_context):
    client = speculative_context["client"]
    speculative_context["response_b"] = client.execute_command(
        "orders", "order-010", "CommandB"
    )


@then("each speculation should start from the same base state")
def then_each_starts_from_base(speculative_context):
    assert speculative_context.get("response_a") is not None
    assert speculative_context.get("response_b") is not None


@then("results should be independent")
def then_results_independent(speculative_context):
    pass  # Each speculative execution is independent


# ==========================================================================
# Error Handling
# ==========================================================================


@given("the speculative service is unavailable")
def given_speculative_service_unavailable(speculative_context):
    speculative_context["client"] = MockSpeculativeClient()
    speculative_context["client"].available = False


@when("I attempt speculative execution")
def when_attempt_speculative_execution(speculative_context):
    try:
        client = speculative_context["client"]
        client.execute_command("orders", "order-001", "TestCommand")
    except ConnectionError as e:
        speculative_context["error"] = e


@then("the speculative operation should fail with connection error")
def then_speculative_operation_should_fail_with_connection_error(speculative_context):
    error = speculative_context.get("error")
    assert error is not None, "Expected connection error"
    assert "connection" in str(error).lower() or isinstance(error, ConnectionError)


@when("I attempt speculative execution with missing parameters")
def when_attempt_with_missing_params(speculative_context):
    speculative_context["error"] = ValueError("Missing required parameters")


@then("the speculative operation should fail with invalid argument error")
def then_speculative_operation_should_fail_with_invalid_argument_error(
    speculative_context,
):
    error = speculative_context.get("error")
    assert error is not None, "Expected invalid argument error"

File: client/rust/tests/steps/speculative_client.rs
//! SpeculativeClient step definitions.

use cucumber::{given, then, when, World};
use std::collections::HashMap;

/// Mock event for testing.
#[derive(Debug, Clone)]
struct MockEvent {
    #[allow(dead_code)]
    sequence: u32,
    #[allow(dead_code)]
    event_type: String,
}

/// Mock EventBook for testing.
#[derive(Debug, Clone, Default)]
struct MockEventBook {
    #[allow(dead_code)]
    domain: String,
    #[allow(dead_code)]
    root: String,
    #[allow(dead_code)]
    events: Vec<MockEvent>,
}

/// Speculative result.
#[derive(Debug, Clone)]
struct SpeculativeResult {
    events: Vec<MockEvent>,
    commands: Vec<String>,
    projection: Option<String>,
    rejection: Option<String>,
}

/// Test context for SpeculativeClient scenarios.
#[derive(Debug, Default, World)]
pub struct SpeculativeClientWorld {
    client_connected: bool,
    aggregates: HashMap<String, MockEventBook>,
    speculative_result: Option<SpeculativeResult>,
    events_persisted: bool,
    edition_created: bool,
    edition_discarded: bool,
    error: Option<String>,
    error_type: Option<String>,
    service_available: bool,
    aggregate_state: String,
    saga_origin: Option<String>,
    has_correlation_id: bool,
    spec_a_result: Option<SpeculativeResult>,
    spec_b_result: Option<SpeculativeResult>,
    real_event_count: u32,
}

// ==========================================================================
// Background Steps
// ==========================================================================

#[given("a SpeculativeClient connected to the test backend")]
async fn given_speculative_client(world: &mut SpeculativeClientWorld) {
    world.client_connected = true;
    world.service_available = true;
}

// ==========================================================================
// Given Steps
// ==========================================================================

#[given(expr = "an aggregate {string} with root {string} has {int} events")]
async fn given_aggregate_with_events(
    world: &mut SpeculativeClientWorld,
    domain: String,
    root: String,
    count: u32,
) {
    let key = format!("{}:{}", domain, root);
    let mut events = vec![];
    for i in 0..count {
        events.push(MockEvent {
            sequence: i,
            event_type: "Event".to_string(),
        });
    }
    world.aggregates.insert(
        key,
        MockEventBook {
            domain,
            root,
            events,
        },
    );
    world.real_event_count = count;
}

#[given(expr = "an aggregate {string} with root {string} in state {string}")]
async fn given_aggregate_in_state(
    world: &mut SpeculativeClientWorld,
    domain: String,
    root: String,
    state: String,
) {
    let key = format!("{}:{}", domain, root);
    world.aggregates.insert(
        key,
        MockEventBook {
            domain,
            root,
            events: vec![MockEvent {
                sequence: 0,
                event_type: "StateChanged".to_string(),
            }],
        },
    );
    world.aggregate_state = state;
}

#[given(expr = "an aggregate {string} with root {string}")]
async fn given_aggregate(world: &mut SpeculativeClientWorld, domain: String, root: String) {
    let key = format!("{}:{}", domain, root);
    world.aggregates.insert(
        key,
        MockEventBook {
            domain,
            root,
            events: vec![],
        },
    );
}

#[given(expr = "events for {string} root {string}")]
async fn given_events_for(world: &mut SpeculativeClientWorld, domain: String, root: String) {
    let key = format!("{}:{}", domain, root);
    world.aggregates.insert(
        key,
        MockEventBook {
            domain,
            root,
            events: vec![MockEvent {
                sequence: 0,
                event_type: "Event".to_string(),
            }],
        },
    );
}

#[given(expr = "{int} events for {string} root {string}")]
async fn given_n_events_for(
    world: &mut SpeculativeClientWorld,
    count: u32,
    domain: String,
    root: String,
) {
    let key = format!("{}:{}", domain, root);
    let mut events = vec![];
    for i in 0..count {
        events.push(MockEvent {
            sequence: i,
            event_type: "Event".to_string(),
        });
    }
    world.aggregates.insert(
        key,
        MockEventBook {
            domain,
            root,
            events,
        },
    );
}

#[given("events with saga origin from \"inventory\" aggregate")]
async fn given_events_with_saga_origin(world: &mut SpeculativeClientWorld) {
    world.saga_origin = Some("inventory".to_string());
}

#[given("correlated events from multiple domains")]
async fn given_correlated_events(world: &mut SpeculativeClientWorld) {
    world.has_correlation_id = true;
}

#[given("events without correlation ID")]
async fn given_events_without_correlation(world: &mut SpeculativeClientWorld) {
    world.has_correlation_id = false;
}

#[given(expr = "a speculative aggregate {string} with root {string} has {int} events")]
async fn given_speculative_aggregate(
    world: &mut SpeculativeClientWorld,
    domain: String,
    root: String,
    count: u32,
) {
    let key = format!("{}:{}", domain, root);
    let mut events = vec![];
    for i in 0..count {
        events.push(MockEvent {
            sequence: i,
            event_type: "Event".to_string(),
        });
    }
    world.aggregates.insert(
        key,
        MockEventBook {
            domain,
            root,
            events,
        },
    );
    world.real_event_count = count;
}

#[given("the speculative service is unavailable")]
async fn given_service_unavailable(world: &mut SpeculativeClientWorld) {
    world.service_available = false;
}

// ==========================================================================
// When Steps
// ==========================================================================

#[when(expr = "I speculatively execute a command against {string} root {string}")]
async fn when_speculative_execute(
    world: &mut SpeculativeClientWorld,
    _domain: String,
    _root: String,
) {
    world.speculative_result = Some(SpeculativeResult {
        events: vec![MockEvent {
            sequence: 0,
            event_type: "SpeculativeEvent".to_string(),
        }],
        commands: vec![],
        projection: None,
        rejection: None,
    });
    world.events_persisted = false;
    world.edition_created = true;
    world.edition_discarded = true;
}

#[when(expr = "I speculatively execute a command as of sequence {int}")]
async fn when_speculative_as_of_sequence(world: &mut SpeculativeClientWorld, _seq: u32) {
    world.speculative_result = Some(SpeculativeResult {
        events: vec![MockEvent {
            sequence: 0,
            event_type: "SpeculativeEvent".to_string(),
        }],
        commands: vec![],
        projection: None,
        rejection: None,
    });
    world.events_persisted = false;
}

#[when(expr = "I speculatively execute a {string} command")]
async fn when_speculative_execute_command(world: &mut SpeculativeClientWorld, cmd_type: String) {
    if cmd_type == "CancelOrder" && world.aggregate_state == "shipped" {
        world.speculative_result = Some(SpeculativeResult {
            events: vec![],
            commands: vec![],
            projection: None,
            rejection: Some("cannot cancel shipped order".to_string()),
        });
    } else {
        world.speculative_result = Some(SpeculativeResult {
            events: vec![MockEvent {
                sequence: 0,
                event_type: format!("{}Executed", cmd_type),
            }],
            commands: vec![],
            projection: None,
            rejection: None,
        });
    }
    world.events_persisted = false;
}

#[when("I speculatively execute a command with invalid payload")]
async fn when_speculative_invalid_payload(world: &mut SpeculativeClientWorld) {
    world.error = Some("Validation error".to_string());
    world.error_type = Some("validation".to_string());
    world.speculative_result = None;
}

#[when("I speculatively execute a command")]
async fn when_speculative_execute_simple(world: &mut SpeculativeClientWorld) {
    world.speculative_result = Some(SpeculativeResult {
        events: vec![MockEvent {
            sequence: 0,
            event_type: "SpeculativeEvent".to_string(),
        }],
        commands: vec![],
        projection: None,
        rejection: None,
    });
    world.events_persisted = false;
    world.edition_created = true;
    world.edition_discarded = true;
}

#[when(expr = "I speculatively execute projector {string} against those events")]
async fn when_speculative_projector(world: &mut SpeculativeClientWorld, projector: String) {
    world.speculative_result = Some(SpeculativeResult {
        events: vec![],
        commands: vec![],
        projection: Some(format!("{} projection result", projector)),
        rejection: None,
    });
    world.events_persisted = false;
}

#[when(expr = "I speculatively execute projector {string}")]
async fn when_speculative_projector_simple(world: &mut SpeculativeClientWorld, projector: String) {
    world.speculative_result = Some(SpeculativeResult {
        events: vec![],
        commands: vec![],
        projection: Some(format!("{} projection result", projector)),
        rejection: None,
    });
}

#[when(expr = "I speculatively execute saga {string}")]
async fn when_speculative_saga(world: &mut SpeculativeClientWorld, _saga: String) {
    world.speculative_result = Some(SpeculativeResult {
        events: vec![],
        commands: vec!["SagaCommand1".to_string(), "SagaCommand2".to_string()],
        projection: None,
        rejection: None,
    });
    world.events_persisted = false;
}

#[when(expr = "I speculatively execute process manager {string}")]
async fn when_speculative_pm(world: &mut SpeculativeClientWorld, _pm: String) {
    if !world.has_correlation_id {
        world.error = Some("Missing correlation ID".to_string());
        world.error_type = Some("missing_correlation".to_string());
        return;
    }

    world.speculative_result = Some(SpeculativeResult {
        events: vec![],
        commands: vec!["PMCommand1".to_string()],
        projection: None,
        rejection: None,
    });
    world.events_persisted = false;
}

#[when("I speculatively execute a command producing 2 events")]
async fn when_speculative_multi_event(world: &mut SpeculativeClientWorld) {
    world.speculative_result = Some(SpeculativeResult {
        events: vec![
            MockEvent {
                sequence: 0,
                event_type: "Event1".to_string(),
            },
            MockEvent {
                sequence: 1,
                event_type: "Event2".to_string(),
            },
        ],
        commands: vec![],
        projection: None,
        rejection: None,
    });
    world.events_persisted = false;
}

#[when(expr = "I verify the real events for {string} root {string}")]
async fn when_verify_real_events(
    _world: &mut SpeculativeClientWorld,
    _domain: String,
    _root: String,
) {
    // Real events remain unchanged
}

#[when("I speculatively execute command A")]
async fn when_speculative_command_a(world: &mut SpeculativeClientWorld) {
    world.spec_a_result = Some(SpeculativeResult {
        events: vec![MockEvent {
            sequence: 0,
            event_type: "EventA".to_string(),
        }],
        commands: vec![],
        projection: None,
        rejection: None,
    });
}

#[when("I speculatively execute command B")]
async fn when_speculative_command_b(world: &mut SpeculativeClientWorld) {
    world.spec_b_result = Some(SpeculativeResult {
        events: vec![MockEvent {
            sequence: 0,
            event_type: "EventB".to_string(),
        }],
        commands: vec![],
        projection: None,
        rejection: None,
    });
}

#[when("I attempt speculative execution")]
async fn when_attempt_speculative(world: &mut SpeculativeClientWorld) {
    if !world.service_available {
        world.error = Some("Connection error".to_string());
        world.error_type = Some("connection".to_string());
    }
}

#[when("I attempt speculative execution with missing parameters")]
async fn when_attempt_missing_params(world: &mut SpeculativeClientWorld) {
    world.error = Some("Invalid argument".to_string());
    world.error_type = Some("invalid_argument".to_string());
}

// ==========================================================================
// Then Steps
// ==========================================================================

#[then("the response should contain the projected events")]
async fn then_response_contains_events(world: &mut SpeculativeClientWorld) {
    let result = world
        .speculative_result
        .as_ref()
        .expect("Should have result");
    assert!(!result.events.is_empty());
}

#[then("the events should NOT be persisted")]
async fn then_events_not_persisted(world: &mut SpeculativeClientWorld) {
    assert!(!world.events_persisted);
}

#[then("the command should execute against the historical state")]
async fn then_execute_against_historical(_world: &mut SpeculativeClientWorld) {
    // Verified by as_of_sequence
}

#[then(expr = "the response should reflect state at sequence {int}")]
async fn then_response_reflects_state(world: &mut SpeculativeClientWorld, _seq: u32) {
    assert!(world.speculative_result.is_some());
}

#[then("the response should indicate rejection")]
async fn then_response_rejection(world: &mut SpeculativeClientWorld) {
    let result = world
        .speculative_result
        .as_ref()
        .expect("Should have result");
    assert!(result.rejection.is_some());
}

#[then(expr = "the rejection reason should be {string}")]
async fn then_rejection_reason(world: &mut SpeculativeClientWorld, reason: String) {
    let result = world
        .speculative_result
        .as_ref()
        .expect("Should have result");
    let rejection = result.rejection.as_ref().expect("Should have rejection");
    assert!(rejection.contains(&reason));
}

#[then("the operation should fail with validation error")]
async fn then_fail_validation(world: &mut SpeculativeClientWorld) {
    assert_eq!(world.error_type, Some("validation".to_string()));
}

#[then("no events should be produced")]
async fn then_no_events_produced(world: &mut SpeculativeClientWorld) {
    assert!(
        world.speculative_result.is_none()
            || world.speculative_result.as_ref().unwrap().events.is_empty()
    );
}

#[then("an edition should be created for the speculation")]
async fn then_edition_created(world: &mut SpeculativeClientWorld) {
    assert!(world.edition_created);
}

#[then("the edition should be discarded after execution")]
async fn then_edition_discarded(world: &mut SpeculativeClientWorld) {
    assert!(world.edition_discarded);
}

#[then("the response should contain the projection")]
async fn then_response_contains_projection(world: &mut SpeculativeClientWorld) {
    let result = world
        .speculative_result
        .as_ref()
        .expect("Should have result");
    assert!(result.projection.is_some());
}

#[then("no external systems should be updated")]
async fn then_no_external_updates(world: &mut SpeculativeClientWorld) {
    assert!(!world.events_persisted);
}

#[then(expr = "the projector should process all {int} events in order")]
async fn then_projector_processes_all(_world: &mut SpeculativeClientWorld, _count: u32) {
    // Verified by design
}

#[then("the final projection state should be returned")]
async fn then_final_projection_state(world: &mut SpeculativeClientWorld) {
    let result = world
        .speculative_result
        .as_ref()
        .expect("Should have result");
    assert!(result.projection.is_some());
}

#[then("the response should contain the commands the saga would emit")]
async fn then_response_contains_commands(world: &mut SpeculativeClientWorld) {
    let result = world
        .speculative_result
        .as_ref()
        .expect("Should have result");
    assert!(!result.commands.is_empty());
}

#[then("the commands should NOT be sent to the target domain")]
async fn then_commands_not_sent(world: &mut SpeculativeClientWorld) {
    assert!(!world.events_persisted);
}

#[then("the response should preserve the saga origin chain")]
async fn then_preserve_saga_origin(world: &mut SpeculativeClientWorld) {
    assert!(world.saga_origin.is_some());
}

#[then("the response should contain the PM's command decisions")]
async fn then_response_contains_pm_commands(world: &mut SpeculativeClientWorld) {
    let result = world
        .speculative_result
        .as_ref()
        .expect("Should have result");
    assert!(!result.commands.is_empty());
}

#[then("the commands should NOT be executed")]
async fn then_commands_not_executed(world: &mut SpeculativeClientWorld) {
    assert!(!world.events_persisted);
}

#[then("the speculative PM operation should fail")]
async fn then_pm_operation_fails(world: &mut SpeculativeClientWorld) {
    assert!(world.error.is_some());
}

#[then("the error should indicate missing correlation ID")]
async fn then_error_missing_correlation(world: &mut SpeculativeClientWorld) {
    assert_eq!(world.error_type, Some("missing_correlation".to_string()));
}

#[then(expr = "I should receive only {int} events")]
async fn then_receive_only_n_events(world: &mut SpeculativeClientWorld, count: u32) {
    assert_eq!(world.real_event_count, count);
}

#[then("the speculative events should not be present")]
async fn then_speculative_not_present(world: &mut SpeculativeClientWorld) {
    // Speculative events are not persisted
    assert!(!world.events_persisted);
}

#[then("each speculation should start from the same base state")]
async fn then_same_base_state(world: &mut SpeculativeClientWorld) {
    assert!(world.spec_a_result.is_some());
    assert!(world.spec_b_result.is_some());
}

#[then("results should be independent")]
async fn then_results_independent(world: &mut SpeculativeClientWorld) {
    let a = world.spec_a_result.as_ref().expect("Should have A result");
    let b = world.spec_b_result.as_ref().expect("Should have B result");
    // Different event types
    assert_ne!(a.events[0].event_type, b.events[0].event_type);
}

#[then("the speculative operation should fail with connection error")]
async fn then_fail_connection(world: &mut SpeculativeClientWorld) {
    assert_eq!(world.error_type, Some("connection".to_string()));
}

#[then("the speculative operation should fail with invalid argument error")]
async fn then_fail_invalid_argument(world: &mut SpeculativeClientWorld) {
    assert_eq!(world.error_type, Some("invalid_argument".to_string()));
}
